src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -1888,22 +1972,40 @@ private static BigInteger Subtract(uint[]? leftBits, int leftSign, uint[]? right
 
         public static BigInteger operator <<(BigInteger value, int shift)
         {
-            if (shift == 0) return value;
-            else if (shift == int.MinValue) return ((value >> int.MaxValue) >> 1);
-            else if (shift < 0) return value >> -shift;
+            if (shift == 0)
+                return value;
+
+            if (shift == int.MinValue)
+                return (value >> int.MaxValue) >> 1;
 
-            int digitShift = shift / kcbitUint;
-            int smallShift = shift - (digitShift * kcbitUint);
+            if (shift < 0)
+                return value >> -shift;
 
-            uint[] xd; int xl; bool negx;
-            negx = GetPartsForBitManipulation(ref value, out xd, out xl);
+            (int digitShift, int smallShift) = Math.DivRem(shift, kcbitUint);
 
-            int zl = xl + digitShift + 1;
-            uint[] zd = new uint[zl];
+            Span<uint> xd = stackalloc uint[1];
+            bool negx = GetPartsForBitManipulation(ref value, ref xd);
+
+            int zl = xd.Length + digitShift + 1;
+            uint[]? zdArray = null;
+            Span<uint> zd = stackalloc uint[0];
+            if (zl <= StackallocUInt32Limit)
+            {
+                zd = stackalloc uint[StackallocUInt32Limit].Slice(0, zl);
+                if (digitShift != 0)
+                {
+                    zd.Slice(0, digitShift).Clear();
+                    zd[^1] = 0;
+                }
```

Comment:
Is the conditional false enough to warrant it?  If yes, is the partial write followed by the next write actually faster than just doing `zd.Clear()`?

Reply:
> Is the conditional false enough to warrant it?

Not sure.  That would mean a `<< shift` with a shift value >= 32, right?  I don't know all the use cases for BigInteger to know how common that is, and unfortunately I've not figured out how to get source.dot.net to show me all call sites for an operator.  (All of my efforts here started because of the >> usage within cert creation, and those methods popping in profiles.)

> If yes, is the partial write followed by the next write actually faster than just doing zd.Clear()?

I think so?  The Clear can be relatively expensive, and it seems likely the shift amount is generally a lot smaller than the BigInteger's bits.